### ISSUE TITLE
Label smoothing for LF-MMI

### ISSUE THREAD

**Issue by danpovey** [k2-fsa/snowfall#179]
Perhaps @zhu-han can try this..
we should be able to implement label smoothing for our LF-MMI system by adding some small constant times `-nnet_output.mean(1).sum() / (len(texts) * accum_grad)` to the loss function [assuming we are still normalizing
by len(texts) which IMO is not as optimal as normalizing by total num-frames, but that's a separate issue.]

**Comment by zhu-han** [comment]
Will try this.

**Comment by zhu-han** [comment]
In `-nnet_output.mean(1)`, is `1` the dimension of output classes?
I did this way by adding 
```
smooth_score = nnet_output.mean(-1).sum()
tot_score += self.smooth_scale * smooth_score
```
in https://github.com/k2-fsa/snowfall/blob/c5ffa3fbaa1bee29b930cc290fcb6a35cb652397/snowfall/objectives/mmi.py#L94
The results are:

smooth scale | test-clean | test-other
-- | -- | --
0 | 6.79 | 17.8
0.01 | 6.8 | 18.5
0.001 | 6.94 | 17.97
0.0001 | 6.73 | 18.05
0.00001 | 6.86 | 17.66

There is no clear improvement by now.

**Comment by csukuangfj** [comment]
If the shape of nnet_output is (N, T, C), should it be
```python
-nnet_output.mean(2).sum()
```
?
It assigns an equal probability for each class per frame and takes the sum of the weighted log_probs per frame,
which is `mean(2)`.

---

Also, the denominator `(len(texts) * accum_grad)` seems to be missing.

**Comment by zhu-han** [comment]
Yes, the shape is (N, T, C). I think `mean(2)` and `mean(-1)` is the same here. I added this `smooth_score` to the `tot_score`. So it will be  normalized by `(len(texts) * accum_grad)` together with original `tot_score`.

**Comment by danpovey** [comment]
Hm.
Try 0.1, just to verify that it gets worse.  If it does, we'll forget this.
The 18.5 is still too close to the margin of error.
And how much is the printed objective function affected by this?

**Comment by zhu-han** [comment]
Will try it.
If we print the original mmi loss (i.e. the total loss - weighted smooth loss) like before as the objf, the validation average objf is not affected much. Take the last validation result of training for example：

smooth scale | validation average objf
-- | --
0 | 0.204217
0.01 | 0.199484
0.001 | 0.217441
0.0001 | 0.20684
0.00001 | 0.206871
With smooth scale 0.01, the validation objf is actually better.

As for the loss value, with smooth scale 0.01, the weighted smooth loss is about 30% ~ 150% of the original mmi loss (i.e. the total loss - weighted smooth loss).

**Comment by danpovey** [comment]
So does the validation loss include that extra term, or you disable it for
validation?

**Comment by zhu-han** [comment]
I disabled it for validation.

**Comment by zhu-han** [comment]
There are results with smooth scale 1 and 0.1.

smooth scale | test-clean | test-other
-- | -- | --
0 | 6.79 | 17.8
1 | 11.42 | 27.1
0.1 | 7.63 | 19.71
0.01 | 6.8 | 18.5

The results are clearly worse than the one without smooth loss.

**Comment by pzelasko** [comment]
Since we have an ali model then maybe another option is to added frame wise cross entropy loss using that alignment, and apply the label smoothing there?

**Comment by danpovey** [comment]
We previously tried regularizing with a cross-entropy loss based on the
alignments of the currently-being-trained model, but we didn't see any
improvements.
Could try it again, of course.
It's possible that the issue is, we are using relatively small models and
the limiting factor is learning, not generalization.

**Comment by danpovey** [comment]
https://arxiv.org/pdf/2005.09150v2.pdf

**Comment by zhu-han** [comment]
I'll try it.

**Comment by danpovey** [comment]
thanks!
